### DISCUSSION TITLE
23. August 2022

### DISCUSSION THREAD

**Discussion by dimakuv** [gramineproject/gramine#848] [Meeting notes]
### Agenda

*(please write your proposed agenda items in comments under this discussion)*

- [x] Vijay: Local Attestation and UNIX Domain Sockets, see https://github.com/gramineproject/gramine/issues/814

The whole meeting was spent on discussing the above item from Vijay. The summary of the discussions can be found in #814. See https://github.com/gramineproject/gramine/issues/814#issuecomment-1226975883

**Comment by vans163** [comment]
![image](https://user-images.githubusercontent.com/3028982/185485447-32368f42-2317-4602-bb47-c05545d4aecf.png)
Can anyone go a bit indepth on these 3rd gen Xeon bios options?

- SGX Package Info In-Band access (?)
- PRMRR Size (this is just mem reserved to SGX at boot, not the max mem per enclave?)
- SGX QoS (?)
- Epoch 1/2 (extra entropy per enclave key?)
- PUBKEYHASH0-4 (?)
- SGX Auto MP Registration (?)

**Reply by dimakuv** [reply]
@vans163 Feels like the most inappropriate place to have this conversation, but here we go.

Most of my answers are based on the information kindly provided by @ScottR-Intel. Also see the replies from @jsun39 below.

> - SGX Package Info In-Band access

This enables publishing of the platform's key blobs via UEFI variables. This is useful to save off the key blobs in case they are deleted (ie SGX Factory reset or the FLASH containing them are erased). More info in our [MP SW Guide](https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_DCAP_Multipackage_SW.pdf).

What are *key blobs*? From the document linked above: "Data Structure that stores the platform keys for each CPU device on the platform. Each device uses a unique sealing key to encrypt the platform keys in the key blob. Used by BIOS to determine the state of the platform keys."

> PRMRR Size

This is how much RAM is dedicated to the SGX-special memory. PRM stands for Processor Reserved memory. It contains some SGX metadata memory, as well as the EPC (Enclave Page Cache).

The best description you can find is in the [Intel SGX Explained paper](https://eprint.iacr.org/2016/086.pdf). Check Section 5.1 and also search for the `PRMRR` keyword in that document. Figure 60 explains it perfectly.

I am unsure what you mean by "not the max mem per enclave" phrase. You seem to refer here to the *virtual enclave address space*, whereas PRMRR describes the *physical enclave address space*. Please note that *virtual enclave size* is basically unrestricted, but all enclaves' memory on a single machine are backed by the PRM/EPC physical address range (multiplexing of seemingly-unlimited virtual memory onto a limited range of physical memory).

> SGX QoS

This is an SGX performance enhancer. Unfortunately, I am not allowed to say more about this feature currently.

> Epoch 1/2

This is otherwise known as OWNER EPOCH (or OWNEREPOCH, written in one word). This is simply another input to the platform's key derivations. The idea was that a platform owner could set these (manually or randomly) and then if they ever retire or give away the platform, all they need to do is change these and data sealed previously on that platform before could not be unsealed by subsequent owners of the platform.

The best description you can find is in the [Intel SGX Explained paper](https://eprint.iacr.org/2016/086.pdf). Search for `OWNEREPOCH` keyword.

> PUBKEYHASH0-3

This was introduced for the Flexible Launch Control (FLC, or just Launch Control) SGX hardware feature. These registers contain the Launch Enclave Public Key Hash (i.e., the hash of the public key of the SGX-special Launch Enclave, or LE).

The story of FLC is long and winded, I suggest to read some [Linux-kernel documentation](https://lore.kernel.org/lkml/20191129231326.18076-24-jarkko.sakkinen@linux.intel.com/). Sorry, was lazy to find a better link, but check the Section "Launch Control". For how exactly these 4 registers correspond to the public key hash, see e.g. [this web page](https://techlibrary.hpe.com/docs/iss/proliant-gen10-uefi/GUID-610261F0-3E25-4E90-8906-297D5E77E521.html).

If I remember correctly, the Linux kernel always assumes that these PUBKEYHASH registers are "unlocked" (you can see it in your BIOS as `Write Enable`), and the Linux kernel just overwrites these registers every time it starts a new enclave (with the MRSIGNER of the new enclave), just to make the SGX hardware happy (the SGX hardware verifies PUBKEYHASH registers against the MRSIGNER on the [`EINIT` instruction](https://www.felixcloutier.com/x86/einit)). The best source of this trick I could find is described [here](https://lkml.org/lkml/2017/3/8/605).

> SGX Auto MP Registration

This sets a UEFI variable bit telling the SGX [Multi-package Registration Agent (aka MPA)](https://github.com/intel/SGXDataCenterAttestationPrimitives/tree/master/tools/SGXPlatformRegistration), if installed, to automatically try to directly register the platform with Intel Registration Services if it has not been done yet (ie. a platform manifest is available in the UEFI variables). More info on this is also in the [MP SW Guide](https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_DCAP_Multipackage_SW.pdf).

(Few! This was quite a long post. I wonder if there should be some proper place where we could collect such bits of knowledge, but I cannot come up with a good proposal.)

**Reply by vans163** [reply]
Thanks this is very good and I thought these weekly meetings were kind of like "office-hours" that's why I asked.  

More about the PRMRR, my understanding is this memory is fully reserved and system cannot use it. So if PRMRR is 2GB and
say 8GB SGX enclave is started+uses_all_8GB or 256MB enclave that grows in size to 8GB+uses_all_8GB, will this fail, as PRMRR is 2GB?
What about if PRMRR is 2GB and 10 enclaves with 256MB are started+use_256MB, will the last few fail?

Is it not possible to use DCAP without registering the platform? I thought the new attestation in SGX2 could be done without intel in the loop (or was planned to).

**Reply by dimakuv** [reply]
> What about if PRMRR is 2GB and 10 enclaves with 256MB are started+use_256MB, will the last few fail?

This depends on the total amount of available physical memory (RAM + swap file), not on the size of PRMRR. As soon as PRM (more specifically, its EPC part that holds the 4KB enclave pages) starts to become full, the Linux kernel starts swapping out the least-recently-used enclave pages out of the EPC and into RAM (and if there is not enough free pages in RAM, then the Linux kernel further tries to swap out from RAM to the swap file on hard drive).

Of course, the swap operation is very slow. So when your SGX enclaves eat up all PRM/EPC memory, this swapping process starts, and the whole system's performance drops. I suggest to also read our explanation here: https://gramine.readthedocs.io/en/latest/manifest-syntax.html#enclave-size

But yes, if you start 10 256MB enclaves (i.e., 2GB of enclave size in total) and your system has only 2GB of *RAM + swap file*, then you're doomed -- Linux has no other choice but to kill one of these SGX enclaves.

> Is it not possible to use DCAP without registering the platform? I thought the new attestation in SGX2 could be done without intel in the loop (or was planned to).

Hm, I'm not sure about this question. I hope @ScottR-Intel can reply here.

**Reply by vans163** [reply]
So system has 256GB of ram in these examples (128GB ram not feasible for 3rd Gen Xeon because you need to populate atleast 8 dimm slots, and 16GB sticks are only about 30-50% cheaper than 32GB sticks [not the usual ~90% cheaper due to 2x less density]). 

Swap is disabled, 0 bytes available for swap.

Keeping PRMRR at 2GB should be fine then, even if starting enclaves that will eventually physically consume 32G of ram?

**Reply by dimakuv** [reply]
> Keeping PRMRR at 2GB should be fine then, even if starting enclaves that will eventually physically consume 32G of ram?

It will be fine from the functionality perspective (SGX enclaves will run correctly, and Linux's OOM killer won't destroy these enclaves due to insufficient physical memory).

But it will be bad from the performance perspective. 32GB of enclaves backed by 2GB of PRMRR will lead to a lot of swap-from-EPC-to-RAM operations, and these operations are expensive. So the performance of your enclave execution will degrade significantly.

TLDR: It depends on your needs. If you are Ok to lose some enclave performance when your enclaves start to exceed 2GB, then keep PRMRR=2GB (and you'll have a lot of RAM available for other, non-SGX-enclave workloads). If the only thing that you plan to run on this system *are* SGX enclaves, then it makes sense to maximize your PRMRR setting.

**Reply by ScottR-Intel** [reply]
Hi @vans163.

First, a little terminology clarification...  In the [Intel SDM](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html), the **only** feature the term "SGX2" refers to is the Enhanced Dynamic Memory Management (EDMM) feature.  It does not refer to any other features like flexible launch control, key separation and sharing, DCAP attestation support, etc.

As for your DCAP question...  On 3rd Gen Xeon Scalable (Ice Lake Server) and later platforms, you will always have to "register" the CPUs/platform with [Intel Registration Service](https://api.portal.trustedservices.intel.com/registration) to be able to generate and verify quotes using ECDSA based SGX remote attestation (RA).  This is a hard requirement and RA will not work without this step.  This can be done in an air-gapped environment if need be, though, utilizing our [PCK Cert ID Retrieval](https://github.com/intel/SGXDataCenterAttestationPrimitives/tree/master/tools/PCKRetrievalTool)  on the air-gapped platform and a simple curl command (see the API docs at the [Intel Registration Service](https://api.portal.trustedservices.intel.com/registration) page) on a platform that is connected to the internet.  You will also, even in an air-gapped situation, have to periodically return to [Intel's Provisioning Certification Services](https://api.portal.trustedservices.intel.com/provisioning-certification) to download updated evidence required to verify quotes.  As an FYI, neither Xeon-E based server nor Core consumer based platforms with SGX support have to do this registration...  only the Scalable CPUs do.

Hope this helps clarify.

Scott

**Reply by vans163** [reply]
@dimakuv Thanks understood.
@ScottR-Intel Thanks understood.

**Comment by jsun39** [comment]
SGX Package Info In-Band access (?)
**When you enable this knob,  you can get the the UEFI variable file: /sys/firmware/efi/efivars/SgxRegistrationPackageInfo-ac406deb-ab92-42d6-aff7-0d78e0826c68. It will contain the per-package key blob information, by default this information is saved in NVRAM, when you enable this knob, you can access it OS. Of course, the context is encrypted.
PRMRR Size (this is just mem reserved to SGX at boot, not the max mem per enclave?)
** No, it means the whole EPC size.
SGX QoS (?)
** SGX Quality of service
Epoch 1/2 (extra entropy per enclave key?)
** different EPOCH will result different sgx key. So you can set your owner EPOCH value, when you want to retire this platform or transfer this platform to others, to protect your data, you can just change the EPOCH value, then others could NOT decrypt your data.
PUBKEYHASH0-4 (?)
** it will be filled with the loaded enclave's MRsigner.  By the way, it should be 0~3, not 0~4.
SGX Auto MP Registration (?)
** this know will enable the automatic registration or not. When this knob is enable, you can install Multi-package Agent, then when system boot into OS, the Multi-package agent server will automatically register your platform to Intel Registration Server.
